pkg/stanza/fileconsumer/internal/trie/trie_test.go

```diff
@@ -0,0 +1,561 @@
+					value:          []byte("ABCD"),
+					delete:         true,
+					deleteExpected: true,
+				},
+				{
+					value:         []byte("ABCD"),
+					matchExpected: false,
+				},
+				{
+					value:         []byte("ABCDEF"),
+					matchExpected: true,
+				},
+				{
+					value:         []byte("ABCDEFGHI"),
+					matchExpected: true,
+				},
+			},
+		},
+		{
+			name:         "TrieCase_Different",
+			initialItems: []string{"ABCD", "XYZ"},
+			testCases: []testCase{
+				{
+					value: []byte("ABCEFG"),
+				},
+				{
+					value:         []byte("ABCDXYZ"),
+					matchExpected: true,
+				},
+				{
+					value: []byte("ABXE"),
+				},
+			},
+		},
+		{
+			name:         "TrieCase_Exact",
+			initialItems: []string{"ABCDEFG", "ABCD"},
+			testCases: []testCase{
+				{
+					value:         []byte("ABCDEFG"),
+					matchExpected: true,
+				},
+				{
+					value:         []byte("ABCD"),
+					matchExpected: true,
+				},
+				{
+					value:         []byte("ABCDE"),
+					matchExpected: true,
+				},
+			},
+		},
+		{
+			name:         "TrieCase_DeleteFalse",
+			initialItems: []string{"ABCDEFG"},
+			testCases: []testCase{
+				{
+					value:          []byte("ABCDEFG"),
+					delete:         true,
+					deleteExpected: true,
+				},
+				{
+					value: []byte("ABCD"),
+				},
+				{
+					value:  []byte("XYZ"),
+					delete: true,
+					// it should be false, as we haven't inserted such values
+					deleteExpected: false,
+				},
+			},
+		},
+		{
+			name:         "TrieCase_Complex",
+			initialItems: []string{"ABCDE", "ABC"},
+			testCases: []testCase{
+				{
+					value:         []byte("ABCDEXYZ"),
+					matchExpected: true,
+				},
+				{
+					value:         []byte("ABCXYZ"),
+					matchExpected: true,
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		trie := NewTrie()
+		for _, k := range tc.initialItems {
+			trie.Put([]byte(k))
+		}
+		t.Run(tc.name, func(t *testing.T) {
+			for _, T := range tc.testCases {
+				runTest(t, trie, &T)
+			}
+		})
+	}
+}
+
+func runTest(t *testing.T, trie *Trie, tc *testCase) {
+	if tc.delete {
+		// Delete the value and check if it was deleted successfully
+		assert.Equal(t, trie.Delete(tc.value), tc.deleteExpected)
+	} else {
+		assert.Equal(t, trie.HasKey(tc.value), tc.matchExpected)
+		if !tc.matchExpected {
+			trie.Put(tc.value)
+		}
+	}
+}
+
+func TestTrieOpSequences(t *testing.T) {
+	opTree{
+		continuations: map[string]opTree{
+			"Found:ABC": opTree{ // First poll finds only ABC.
+				ops: []testOp{
+					put("ABC"),
+				},
+				continuations: map[string]opTree{
+					"Done:ABC": opTree{ // Finish reading ABC and remove from trie
+						ops: []testOp{
+							del("ABC", true, "was just added"),
+							has("ABC", false, "was just deleted"),
+						},
+					},
+					"Found:ABCDEF": opTree{ // Next poll finds ABCDEF
+						ops: []testOp{
+							has("ABCDEF", true, "recognize ABC w/ DEF appended"),
+							put("ABCDEF"), // TODO HasKey returns true, so how do we know to call this?
```

Comment:
Regarding this TODO, do we need a way to differentiate prefix matches from exact matches? Otherwise I'm not sure we we can take advantage of the "push" functionality in `Put`.

We could switch back to `Get` and return the value that "matched" the key. The caller can interpret whether or not it was what they passed as a parameter or just a prefix. WDYT?

Reply:
Yeah we can do that.
Also, If we push the exact matches then it would make no difference to trie.

Reply:
> If we push the exact matches then it would make no difference to trie.

In most situations I'd be ok with duplicating the `Put` but this package is optimized wherever possible so I'd prefer to err on the side of performance here.